peps/pep-9999.rst

```diff
@@ -157,36 +157,406 @@ splitting off headers and other build-time dependencies in a separate
+subcommand in any Python development workflow tool (e.g., Pip, Poetry, Hatch, PDM).
 
 
 Registry design
 ---------------
 
-- One central Python package for the registry tool
-- A plugin design, where system package managers can register their name mapping
-- Name mapping should be in a structured, human-readable file format (TBD:
-  JSON, YAML, or ...)
+The mapping infrastructure should include the following components and properties:
+
+- A central registry for known generic and virtual PURLs.
+- A list of known ecosystems and package managers, where ecosystem maintainers
+  can register their name mapping(s).
+- A standardized schemsa that defines how mappings should be structured.
+- The above documents should be written in a structured, human-readable file format.
+  As long as it can be validated with the corresponding JSON Schemas, they could be
+  delivered as JSON, YAML, TOML and/or others. Authors recommend JSON or YAML.
+- One central Python package for the central registry, list of ecosystems and known
+  mappings.
+
+On the client side, there should be:
+
 - A way for the system to specify a default (e.g., the Python install on Ubuntu
   could register ``apt`` as the default system package manager with the
-  registry tool. It may also be left unspecified.
-- A way for the user to specify the default and/or current system package
-  manager. E.g., a user on Ubuntu may want either ``apt``, ``conda``, ``brew``
-  or ``spack`` as their package manager of choice to provide external
-  dependencies.
+  registry tool). It may also be left unspecified.
+- A way for the user to specify the default and/or current system package manager.
+  E.g., a user on Ubuntu may want either ``apt``, ``conda``, ``brew``  or ``spack``
+  as their package manager of choice to provide external dependencies.
 
 
 Specification
 =============
 
-
+Three schemas are proposed:
+
+1. A central registry of PURLs.
+2. A list of known ecosystems and their package manager names.
+3. The ecosystem-specific mappings of PURLs to package specifiers,
+   plus details of their package managers.
+
+The central registry defines which identifiers are recognized as canonical. Each entry MUST
+provide a valid PURL in the field ``id``, with an optional free form ``description`` text.
+Additionally some entries MAY refer to another entry via the ``provides`` field, which takes a
+list of strings already defined as ``id`` in the registry. This is useful for aliases
+(e.g. ``dep:generic/arrow`` and ``dep:github/apache/arrow``), and concrete implementations of
+a ``dep:virtual/`` entry (e.g. ``dep:generic/gcc`` would provide ``dep:virtual/compiler/c``).
+Entries without ``provides`` content or, if populated, only with ``dep:virtual/`` identifiers,
+are considered canonical.
+
+The list of known ecosystems assigns an identifier to each ecosystem and reports the
+canonical location for its mapping. The known package managers are also provided here
```

Comment:
This canonical location may turn out to be package manager specific, right? Even if it's the same list, it's not clear to me that it makes sense to have a shared location. That may lead to conflicts between updates of different package managers.

Reply:
Not sure what you mean with package manager specific. I've been assuming each _ecosystem_ would have a centralized mapping. As in, it doesn't make sense for `mamba` and `pixi` to maintain different mappings, more like a central conda ecosystem mapping, because they all use the same package names and spec syntax.

Although the current design does allow for a rogue package manager to start its own mapping if they chose to design a different package spec syntax, I guess.

Reply:
> I've been assuming each _ecosystem_ would have a centralized mapping. As in, it doesn't make sense for `mamba` and `pixi` to maintain different mappings, more like a central conda ecosystem mapping

What I'm saying is that your assumption here may not be correct. E.g, `pixi` has its own mapping of PyPI to conda-forge, not shared with other tools, I believe. Similarly, I have no idea whether `yum`/`dnf` are sharing a mapping (or would be willing to). When there are multiple package managers, they tend to compete rather than collaborate closely.

Reply:
Prefix has created their own mapping _schema_ and population method, but the "dots" they are connecting are the same in the end: PyPI names and conda-forge names. Pixi doesn't use different package names than conda, for example.

But I get your point: we don't need to assume a coherent ecosystem with respectfully coexisting package managers. There's a chance they establish different syntaxes for their package specifiers (e.g. Poetry) or rename some packages on the fly (not sure this has happened).

Reply:
Yep. And even without different syntaxes, best not to assume collaboration on a shared file for something this important to package manager functioning. E.g., maybe pixi tweaks their mapping temporarily to avoid a pixi-specific resolver bug. Or they just see better mappings as a competitive advantage. Could be any number of reasons.

Reply:
I removed this part now.